tracer/src/Datadog.Trace/Telemetry/Collectors/DependencyTelemetryCollector.cs

```diff
@@ -43,25 +41,18 @@ internal void AssemblyLoaded(AssemblyName assembly)
                || assemblyName.StartsWith("App_global.asax.", StringComparison.Ordinal)
                || assemblyName.StartsWith("App_Code.", StringComparison.Ordinal)
                || assemblyName.StartsWith("App_WebReferences.", StringComparison.Ordinal)))
-             || (assemblyName.Length == 36
-              && assemblyName[8] == '-'
-              && assemblyName[13] == '-'
-              && assemblyName[18] == '-'
-              && assemblyName[23] == '-'))
+             || IsGuid(assemblyName))
             {
                 return;
             }
 
             var key = new DependencyTelemetryData(name: assemblyName) { Version = assembly.Version?.ToString() };
-            if (_assemblies.TryAdd(key, true))
-            {
-                SetHasChanges();
-            }
+            _assemblies.TryAdd(key, true);
         }
 
         public bool HasChanges()
         {
-            return _hasChangesFlag == 1;
+            return !_assemblies.IsEmpty;
```

Comment:
This looks like [it can sometimes be quite expensive](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1410) (though who knows'if that's the case without testing). Keeping the `_hasChangesFlag` for this check may make sense just in case?

Reply:
I agree with that. It looks like `IsEmpty` is optimized for the "not empty" case, whereas we expect the collection to be empty most of the time.

Reply:
Actually, since we don't use the value stored in the dictionary, it might be better to add the items to a ConcurrentQueue then deduplicate them with a hashset:

```

|                   Method |                  Job |              Runtime |     Mean |    Error |   StdDev |   Median |  Gen 0 |  Gen 1 | Allocated |
|------------------------- |--------------------- |--------------------- |---------:|---------:|---------:|---------:|-------:|-------:|----------:|
| WithConcurrentDictionary |             .NET 6.0 |             .NET 6.0 | 54.56 us | 0.510 us | 0.477 us | 54.36 us | 0.3662 |      - |     30 KB |
|      WithConcurrentQueue |             .NET 6.0 |             .NET 6.0 | 42.95 us | 0.193 us | 0.171 us | 42.90 us | 0.4883 |      - |     40 KB |
| WithConcurrentDictionary |        .NET Core 3.1 |        .NET Core 3.1 | 67.08 us | 1.326 us | 1.902 us | 68.21 us | 0.2441 |      - |     29 KB |
|      WithConcurrentQueue |        .NET Core 3.1 |        .NET Core 3.1 | 56.58 us | 0.138 us | 0.122 us | 56.58 us | 0.4272 |      - |     40 KB |
| WithConcurrentDictionary | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 74.13 us | 0.455 us | 0.380 us | 74.08 us | 5.4932 | 0.3662 |     34 KB |
|      WithConcurrentQueue | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 59.07 us | 0.299 us | 0.265 us | 59.01 us | 3.2349 | 0.1221 |     20 KB |
```

Reply:
So actually, as we wanted to keep the versions already sent, I kept the dictionary and just used the value part to differentiate if an aseembly has been sent to the backend.
Done in d52f09aadf1850367dde0f77fedcca62515b0aa8